Add support for env var kind

lcarva opened the pull request:
* CLOUDBLD-1484

nirzari commented on cachito/web/models.py:
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
+        if not isinstance(name, str):
+            raise ValidationError("The name of environment variables must be a string")
+        if not isinstance(info, dict):
+            raise ValidationError("The info of environment variables must be an object")
```
Do we want to specify `a dict object` instead ?

lcarva replied:
It's a little awkward because this error bubbles up to the API error response which deals in term of Json, not python.

mprahl commented on README.md:
```diff
@@ -190,7 +190,9 @@ Custom configuration for the Celery workers are listed below:
-  set for that package manager and the values are the environment variable values. Check
+  set for that package manager and the values a dictionary with the keys `value` and `kind`. The
+  `value` must be a string which specifies the value of the environment variable. The `kind` must
+  also be a string which specifies the type of value, either `"path"`, or `"string"`. Check
```
```suggestion
  also be a string which specifies the type of value, either `"path"` or `"string"`. Check
```

mprahl commented on cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:
```diff
@@ -0,0 +1,52 @@
+"""Add environment variables types
+
+Revision ID: cb6bdbe533cc
+Revises: 615c19a1cee1
```
FYI, one of us will have to modify this depending on if this PR is merged before #181.

lcarva replied:
Ack.

mprahl commented on cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:
```diff
@@ -0,0 +1,52 @@
+
+
+def upgrade():
+    with op.batch_alter_table("environment_variable", schema=None) as batch_op:
```
```suggestion
    with op.batch_alter_table("environment_variable") as batch_op:
```

mprahl commented on cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:
```diff
@@ -0,0 +1,52 @@
+
+def upgrade():
+    with op.batch_alter_table("environment_variable", schema=None) as batch_op:
+        # Drop previous unique_constraint. Since the constraint was not originally added with a
```
Can you just look at the automatic name used by Postgres and fix the old migration to set the name explicitly instead?

lcarva replied:
IMO that's too brittle. We should instead setup explicit [naming conventions](https://alembic.sqlalchemy.org/en/latest/naming.html), and create another migration script to "rename" them all.

mprahl replied:
We have successfully done this in the past:
* https://github.com/release-engineering/cachito/commit/3bfee393276abc0d5b0c59f18b0340d3129d2689
* https://github.com/release-engineering/cachito/commit/dbb205747d089cfadd4418741a5aa09a1942d778

mprahl commented on cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:
```diff
@@ -0,0 +1,52 @@
+    connection = op.get_bind()
+    connection.execute("UPDATE environment_variable SET kind='path' WHERE kind IS NULL")
+
+    with op.batch_alter_table("environment_variable", schema=None) as batch_op:
```
```suggestion
    with op.batch_alter_table("environment_variable") as batch_op:
```

mprahl commented on cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:
```diff
@@ -0,0 +1,52 @@
+        batch_op.add_column(sa.Column("kind", sa.String(), nullable=True))
+
+    connection = op.get_bind()
+    connection.execute("UPDATE environment_variable SET kind='path' WHERE kind IS NULL")
```
Could you please use SQLAlchemy to do this and not rely on raw SQL?

mprahl commented on cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:
```diff
@@ -0,0 +1,52 @@
+
+
+def downgrade():
+    with op.batch_alter_table("environment_variable", schema=None) as batch_op:
```
```suggestion
    with op.batch_alter_table("environment_variable") as batch_op:
```

mprahl commented on cachito/web/models.py:
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
+        if missing_keys:
+            raise ValidationError(
+                "The following attributes must be set in the info of the environment variables: "
+                f"{', '.join(sorted(missing_keys))}"
```
```suggestion
                ", ".join(sorted(missing_keys))
```

lcarva replied:
This doesn't work. It first concatenates the string, then calls join on it.

mprahl commented on cachito/web/models.py:
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
+        invalid_keys = info.keys() - required_keys
+        if invalid_keys:
+            raise ValidationError(
+                "The following attributes are not allowed in the info of the environment "
```
```suggestion
                "The following keys are not allowed in the info of the environment "
```

lcarva replied:
This has to say attributes because it bubbles up to the API which is JSON based, no?

mprahl replied:
I realize that, but we use the term `keys` everywhere else as far as I know for user facing errors.

lcarva replied:
Alright, I changed this to be consistent with the other instances. 

BTW, we should also change `The following parameters are invalid` to be consistent with this. I can do that in a follow up PR.

mprahl commented on cachito/web/models.py:
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
+        kind = info.get("kind")
+        if not isinstance(kind, str):
+            raise ValidationError("The kind of environment variables must be a string")
+        if kind not in ("path", "string"):
```
Optional: It'd be nice to define the valid kinds at the class level.

mprahl commented on cachito/web/models.py:
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
+        cls.validate_json(name, info)
+        return cls(name=name, **info)
 
     def to_json(self):
```
I don't think this method is used anymore. Could you please delete it if that's the case?

mprahl commented on cachito/web/models.py:
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
+        kind = info.get("kind")
+        if not isinstance(kind, str):
+            raise ValidationError("The kind of environment variables must be a string")
+        if kind not in ("path", "string"):
```
Instead of the kind of `string`, how about `literal`?

lcarva replied:
Good idea. An environment variable must always be a string. Path and literal are just different "types" of strings. Made the change.

mprahl commented on cachito/web/static/api_v1.yaml:
```diff
@@ -360,6 +397,21 @@ components:
+          description: The kind of the environment variable, e.g. "path" or "string"
+      required:
+      - value
+      - kind
```
Could you alphabetize this please?

mprahl commented on cachito/workers/config.py:
```diff
@@ -181,12 +184,23 @@ def validate_celery_config(conf, **kwargs):
+        for env_var_info in pkg_manager_value.values():
+            if not isinstance(env_var_info, dict):
+                raise ConfigError(
+                    'The configuration "cachito_default_environment_variables" must be a '
+                    "dictionary of dictionaries of dictionaries!"
```
Say that five times fast!

lcarva replied:
I did and my dictionary jumped off the shelf.

mprahl commented on cachito/workers/pkg_managers/general.py:
```diff
@@ -111,7 +111,9 @@ def update_request_with_packages(request_id, packages, env_vars=None):
 
     :param list packages: the list of packages that were resolved
-    :param dict env_vars: mapping of environment variables to record
+    :param dict env_vars: mapping of environment variables to record. Attributes represent
```
```suggestion
    :param dict env_vars: mapping of environment variables to record. The keys represent
```

mprahl commented on cachito/workers/tasks/gomod.py:
```diff
@@ -43,7 +43,10 @@ def fetch_gomod_source(request_id, dep_replacements=None):
+        "GOCACHE": {"value": "deps/gomod", "kind": "path"},
+        "GOPATH": {"value": "deps/gomod", "kind": "path"},
+    }
+    env_vars.update(get_worker_config().cachito_default_environment_variables.get("gomod", {}))
```
```suggestion
    env_vars.update(config.cachito_default_environment_variables.get("gomod", {}))
```

mprahl commented on tests/test_api_v1.py:
```diff
@@ -1172,3 +1196,54 @@ def test_request_config_post_not_authorized(auth_env, client, db):
+
+    env_vars_expected = env_vars or {}
+
+    len(EnvironmentVariable.query.all()) == len(env_vars_expected.items())
```
This may work:
```suggestion
    EnvironmentVariable.query.count() == len(env_vars_expected.items())
```

lcarva replied:
It did!

lcarva commented on README.md:
```diff
@@ -190,7 +190,9 @@ Custom configuration for the Celery workers are listed below:
 * `cachito_default_environment_variables` - a dictionary where the keys are names of package
   managers. The values are dictionaries where the keys are default environment variables to
-  set for that package manager and the values are the environment variable values. Check
+  set for that package manager and the values a dictionary with the keys `value` and `kind`. The
```
I went with "and the values are a dictionary"

lcarva commented:
I've addressed all the comments. This is ready for another review round.

mprahl commented on tests/test_api_v1.py:
```diff
@@ -1172,3 +1196,54 @@ def test_request_config_post_not_authorized(auth_env, client, db):
+
+    env_vars_expected = env_vars or {}
+
+    EnvironmentVariable.query.count() == len(env_vars_expected.items())
```
```suggestion
    EnvironmentVariable.query.count() == len(env_vars_expected)
```

mprahl commented on cachito/web/migrations/versions/cdf17fad3edb_request_env_vars.py:
```diff
@@ -23,7 +23,7 @@ def upgrade():
         sa.Column("value", sa.String(), nullable=False),
         sa.PrimaryKeyConstraint("id"),
-        sa.UniqueConstraint("name", "value"),
+        sa.UniqueConstraint("name", "value", name="uq_environmet_variable_name_value"),
```
@lcarva do you mind using the value that Postgres generates by default since we've been using that for other constraints and it'd be nice to be consistent.

lcarva replied:
Sure. What would that be here?

mprahl replied:
```suggestion
        sa.UniqueConstraint("name", "value", name="environment_variable_name_value_key"),
```

lcarva replied:
Done!

lcarva commented:
Rebased, updated the migration script to use the newly added migration script as the `down_revision`.

I also caught a bug in the migration script:
```diff
-        env_var_table.update().where(env_var_table.c.kind is None).values(kind="path")
+        env_var_table.update().where(env_var_table.c.kind == sa.null()).values(kind="path")
```

mprahl commented on cachito/web/migrations/versions/cdf17fad3edb_request_env_vars.py:
```diff
@@ -23,7 +23,7 @@ def upgrade():
         sa.Column("value", sa.String(), nullable=False),
         sa.PrimaryKeyConstraint("id"),
-        sa.UniqueConstraint("name", "value"),
+        sa.UniqueConstraint("name", "value", name="environment_variable_name_value_key"),
```
Optional: This change would be better as a separate commit

mprahl commented on cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:
```diff
@@ -0,0 +1,55 @@
+
+def upgrade():
+    with op.batch_alter_table("environment_variable") as batch_op:
+        # Drop previous unique_constraint. Since the constraint was not originally added with a
```
Please delete this comment

lcarva replied:
Doh! removed it.
